man/openrpg.1
.\"        Title: openrpg
.\"       Author: [see the "Authors" section]
.\"         Date: 12/24/2019
.\"     Language: English

.\" View this man page with the command
.\" man -l openrpg.1
.\" When installing, gzip this file and place it in the man folder

.\" For information on macros used in this file, see man page
.\" groff_man(7)
.\" http://man7.org/linux/man-pages/man7/groff_man.7.html

.\" Header and footer information
.TH "OPENRPG" "1" "2019-12-24" "OPENRPG" "OpenRPG Manual"

.\" NAME section
.\" TODO: Redo this text
.SH "NAME"
openrpg \- emulates ttrpg systems

.\" SYNOPSIS section, includes basic command form
.SH "SYNOPSIS"
.B openrpg
[\fI\,OPTION\/\fR]... <command> [<args>]

.\" DESCRIPTION section, includes command line arguments
.SH "DESCRIPTION"
.PP
.\" TODO: redo this text
OpenRPG allows the user to quickly create characters, generate names, and roll virtual dice from the command line
.PP
Commands and options are split up, as commands are created as different binaries, however some arguments link to different commands. These instances will be marked. 

.\" List of options, alphabetical
.SS "OPTIONS"
.TP
\fB\-h\fR, \fB\-\-help\fR
displays the help message and exits
.TP
\fB\-n\fR, \fB\-\-name\fR=\fI\,RACE\/\fR, \fI\,GENDER\/\fR
links to the name command, launches the subprogram \fBname-generator\fR. See man page \fBopenrpg-name-generator(1)\fR
.TP
\fB\-q\fR, \fB\-\-quiet\fR
don't print the banner on startup
.TP
\fB\-r\fR, \fB\-\-roll\fR=\fI\,XdY\/\fR
links to the roll command, launches the subprogram \fBroll\fR. See man page \fBopenrpg-roll(1)\fR
.TP
\fB\-v\fR, \fB\-\-verbose\fR
.\" TODO: redo this text
supresses quiet flag and sets verbose flag
.TP
\fB\-V\fR, \fB\-\-version\fR
print version info

.\" List of commands, alphabetical
.SS "COMMANDS"
.TP
\fBexit\fR, \fBquit\fR, \fBq\fR
exits the program
.TP
\fBhelp\fR, \fBh\fR \fI\,COMMAND\/\fR
prints help text for the command, if the argument is provided. Otherwise prints the help text for the openrpg program
.TP
\fBgenerate\fR, \fBgen\fR \fI\,RACE\/\fR, \fI\,GENDER\/\fR
launches the subprogram \fBname-generator\fR. See man page \fBopenrpg-name-generator(1)\fR
.TP
\fBroll\fR, \fBr\fR \fI\,XdY\/\fR
launches the subprogram \fBroll\fR. See man page \fBopenrpg-roll(1)\fR
.TP
\fBversion\fR, \fBver\fR, \fBv\fR
print version info
.SH "COPYRIGHT"
.\" TODO copyright text
.\" What's the proper copyright text? For now I'm just putting the license
OpenRPG Software License \- Version 1.0 \- February 10th, 2017 \fB\fI\,https://openrpg.io/about/license/\/\fR

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html). However, note that all pre-v0.5.0 releases will have the suffix "-dev"

## [Unreleased]
### Added
- `man openrpg` thanks to [@var-username](https://github.com/var-username)

### Changed
- Moved the ASSET_LOC from a preprocessor definition to a Core function that searches multiple locations until it finds our data directory, throwing an error if no data can be found
- Core now takes advantage of the STL C++17 Filesystem library to locate our data folder using an array of paths

### Fixed

### Removed

## [0.5.0]
There is likely a TON that has changed that is not noted below. This release is a HUGE one, so we likely missed something. If you notice something is missing, please open a PR to fix it (with commit citations please)!

### Added
- We are now on the Arch User Repository and supporting Arch as our first Linux distro, as it is what I ([@incomingstick](http://github.comincomingstick)) use!
    - You can use the PKGBUILD branch of this Repo to build from `makepkg` that way
    - Clone us out of the AUR yourself by running `git clone https://aur.archlinux.org/openrpg-git.git && cd openrpg-git && makepkg -si`
    - You can use your favorite AUR helper, such as `yay`
        - Run `$ yay -S openrpg`
- We have improved our CMake logic to include the use of CPack for packaging, for all supported OS's! Check out our [downloads](https://openrpg.io/download/) page for your systems installer!
- Build scripts now include a package target for packaging with CPack
- NodeJS and `npm` integrated into the project workflow. You can now use `npm` and subsequently `cmake-js` to compile an additional `orpgAddon.node` binary that is a compiled Node module. 
    - This NodeJS addon aims to be a complete wrap of the C++ libraries. As such, most functions and classes should operate as you would expect the C++ counterpart.
    - Expose the `VERSION` preprocessor definition via the `ORPG_VERSION()` function
- An ASCII character sheet in `data/` thanks to [@var-username](https://github.com/var-username)
    - data/character_sheets/ascii_char_sheet
    - "data/character_sheets/Blank ASCII Character Sheet.txt"
- Added build info for Windows and improved the Unix documentation
- Added an uninstall target to CMake and Make, allowing for the use of `make uninstall` to... uninstall
    - cmake_uninstall.cmake.in
- build.sh has a ton of new commands, including an uninstall command `$ ./build.sh uninstall`
- This Changelog! You are likely looking at CHANGELOG.md
- The OGL is now in our main LICENSE, in addition to the OSL, to hopefully cover our bases with that stuff... hopefully we don't accidentally use copyright WotC stuff, or anyones stuff for that matter
- We now generate export headers via CMake for the Windows DLLs to ensure proper linking
- Some suggested vim settings, if thats your editor of choice
- TravisCI Windows builds! We are now testing Windows via TravisCI, as well as AppVeyor
- OSC XCode8 tests on TravisCI!
- Lots of namespacing to help prevent name collision when our project is used, the structure is as follows:
    - `ORPG`
    - `ORPG::Core`
    - `ORPG::Utils`
    - `ORPG::Roll`
    - `ORPG::Names`
    - `ORPG::Character`
- `NameGenerator` class now has a constructor that takes a file location as a `std::string`, here are some important notes:
    - By creating this constructor we are allowing an end user to specify the location the namelist we will read from. It is extremely important we explore any possible security implications of this further. I [@incomingstick](http://github.comincomingstick)) do think it could prove a beneficial feature to allow end users to specify namelists.
    - I, [@incomingstick](http://github.comincomingstick), created this function specifically to help the test suite. When testing on a fresh system, we are built before we test, but the release builds defines ASSET_LOC as: "/usr/local/data/openrpg" Because of this, it attempts to check a folder that has not been installed yet, and we need a way to point NameGenerator to our source data folder.
- Character module  now takes full advantage of our Names library to allow for the use of first and last names
- Added the Elf race to the Characters library
    - `class Elf : public Character`
    - `class HighElf : public Elf`
- `CharacterFactory` class has a new constructor that takes a name as a `std::string`
- Character library has a new class, `Skill` that helps represent the 5e Skill in a better way, and makes use of the new `EnumSkill` enum to pair score to human readable tags thanks to [@var-username](https://github.com/var-username)
- `ExpressionParser` now handles the modulo operation!

### Changed
- Building Libraries as Dynamic instead of Static, because why wouldn't a game want Dynamic/Shared libraries!
- Split .github/ISSUE_TEMPLATE.md into two files:
    - .github/FEATURE_REQUEST.md
    - .github/BUG_REPORT.md
- Renamed `int random(...)` to `int randomInt(...)` in the Utils namespace thanks to [@var-username](https://github.com/var-username)
- Fixed some install locations to better form to different platforms
- build.bat now has actual commands such as `> .\build.bat clean`, or `> .\build.bat rebuild` as well as better error reporting
- Make sure we don't accidentally add some packaged/compressed files to the repo, cuz that would be obnoxious
- Lots of "auto-fication" across the project, as well as other C++11 improvements
- Lots of restructuring within the files themeselves to help keep things better organized
- Lots of additional documention comments throughout the project for you other devs out there

src/openrpg.cpp
/*
openrpg - openrpg.cpp
Created on: Nov 7, 2016

OpenRPG Software License - Version 1.0 - February 10th, 2017 <https://openrpg.io/about/license/>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
*/
#include <iostream>
#include <vector>
#include <string>

#include "openrpg.h"
#include "names.h"
#include "roll.h"

using namespace std;
using namespace ORPG;

/**
 * TODO hold in memory a history of the commands run in the current session
 * and allow use of the UP and DOWN arrow keys to move through this list
 *
 * NOTE: We should make this finite, based on a changeable setting, up to a max of 1000 lines
 **/
vector<string> commandHistory;

/**
 * @desc This function parses all cla's passed to argv from the command line.
 * This function may terminate the program.
 *
 * @param int argc - the number of arguments passed / the length of argv[]
 * @param char* argv[] - the arguments passed from the command line
 * @return int - an integer code following the C/C++ standard for program success
 **/
int parse_args(int argc, char* argv[]) {
    /* getopt_long stores the option and option index here */
    int opt = 0, opt_ind = 0;

    /* these are the long cla's and their corresponding chars */
    static struct Core::option long_opts[] = {
        {"help",    no_argument,        0,  'h'},
        {"name",    required_argument,  0,  'n'},
        {"quiet",   no_argument,        0,  'q'},
        {"roll",    required_argument,  0,  'r'},
        {"verbose", no_argument,        0,  'v'},
        {"version", no_argument,        0,  'V'},
        /* NULL row to terminate struct */
        {0,         0,                  0,   0}
    };

    while ((opt = getopt_long(argc, argv, "hn:qr:vV", long_opts, &opt_ind))
			!= EOF) {
        switch (opt) {
			/* -h --help */
			case 'h': {
				Core::print_help_flag();
			} break;

			/* -n --name */
			case 'n': {
				NameGenerator name;

				if(Core::optind < argc) {
					if((string)Core::optarg == "male" || (string)Core::optarg == "female") {
                        name.set_gender((string)Core::optarg);
						name.set_race((string)argv[Core::optind++]);
					} else {
						name.set_race((string)Core::optarg);
						name.set_gender((string)argv[Core::optind++]);
					}
				} else if(Core::optind == argc) {
					name.set_race((string)Core::optarg);
				} else {
					fprintf(stderr, "Error: invalid number of args (expects 1 or 2)\n");
					Core::print_help_flag();
				}
				printf("%s\n", name.make_name().c_str());
				exit(EXIT_SUCCESS);
			} break;

			/* -q --quiet */
			case 'q': {
				Core::QUIET_FLAG = true;
				Core::VB_FLAG = false;
			} break;

			/* -r --roll */
			case 'r': {
				ExpressionTree tree;
				tree.set_expression(Core::optarg);
				printf("%i\n", tree.parse_expression());
				exit(EXIT_SUCCESS);
			} break;

			/* -v --verbose */
			case 'v': {
				Core::VB_FLAG = true;
				Core::QUIET_FLAG = false;
			} break;

			/* -V --version */
			case 'V': {
				Core::print_version_flag();
			} break;

			case ':':
			case '?': {
				Core::print_help_flag();
			} break;

			/* if we get here something very bad happened */
			default: {
				fprintf(stderr, "Aborting...\n");
				exit(EXIT_FAILURE);
			}
        }
    }

    return CONTINUE_CODE;
}

/**
 * @desc parse input is our ORPG shell. This takes in the users input string and parses
 * it out based on our logic. This function may terminate the program.
 *
 * @param string in - the users input to be parsed
 * @return int - an integer code following the C/C++ standard for program success
 **/
int parse_input(string in) {
    if (in.size() > 0) {
        // parsed individual words
        vector<string> words;

        // temporary container for word being built
        string word;

       //standardizes inputs to ignore case
        for(int i = 0; (unsigned) i < in.size(); i++) {
            in[i] = tolower(in[i]);

            if((in[i] < '{' && in[i] > '`') || (in[i] < ':' && in[i] > '/') || in[i] == '+' || in[i] == '-') {
                word += in[i]; //pushes character to word
            } else if(word.size() > 0) {
                words.push_back(word); //end of word
                word = {}; //resets word
            }
        }

		//end of command word
        if(word.size() > 0)
			words.push_back(word);

        if (words.size() > 0) {
            // TODO simple commands, must be expanded on based on command content
            // NOTE add character-generator when it's finished
            if(words[0] == "exit" || words[0] == "quit" || words[0] == "q") {
                return EXIT_SUCCESS;
            } else if(words[0] == "generate" || words[0] == "gen" || words[0] == "ng") {
                if(words.size() >= 2) {
                    NameGenerator name;
                    if(words.size() >= 3) {
                        // TODO add nonbinary genders (a la warforged) and/or shortcuts
                        if(words[1] == "male" || words[1] == "female") {
                            name.set_gender(words[1]);
                            name.set_race(words[2]);
                        } else {
                            name.set_race(words[1]);
                            name.set_gender(words[2]);
                        }
                    } else if(words.size() == 2) {
                        name.set_race(words[1]);
                    } else {
                        fprintf(stderr, "Error: invalid number of args (expects 1 or 2)\n");
                    }
                    printf("%s\n", name.make_name().c_str());
                    return CONTINUE_CODE;
                } else {
                    printf("Missing arguments!\n");

                    Names::print_basic_help();

                    return CONTINUE_CODE;
                }
            } else if(words[0] == "roll" || words[0] == "r") {
                // TODO fix the roll command
                if(words.size() > 1) {
                    string exp;
                    for (int i = 1; (unsigned) i < words.size(); i++) {
                        exp += words[i];
                    }
                    ExpressionTree tree;
                    if(tree.set_expression(exp))
                        printf("%i\n", tree.parse_expression());
                    return CONTINUE_CODE;
                } else {
                    printf("Missing arguments\n");

                    Roll::print_basic_help();

                    return CONTINUE_CODE;
                }
            } else if(words[0] == "help" || words[0] == "h" || words[0] == "H") {
                /*
				 * TODO complete the help command as follows
                 *    The help command should print in a similar format
                 *    to the '-h' arguement, but should be expanded on in
                 *    the TUI. Each command will have its own help section
                 *    i.e calling `help roll` will display the roll modules
                 *    TUI formatted help command. This should hold true for
                 *    all modules callable from within the TUI.
                 */
                if(words.size() > 1) {
                    if(words[1] == "roll" || words[1] == "r") {
                        Roll::print_basic_help();

                        return CONTINUE_CODE;
                    } else if(words[1] == "generate" || words[1] == "gen" || words[1] == "ng") {
                        Names::print_basic_help();

                        return CONTINUE_CODE;
                    } else {
                        Core::print_basic_help();

                        return CONTINUE_CODE;
                    }
                } else {
                    Core::print_basic_help();

                    return CONTINUE_CODE;
                }
            } else if (words[0] == "version" || words[0] == "ver" || words[0] == "v" || words[0] == "V") {
                /* Prints print_version_string() without exiting */
                Core::print_basic_version();

                return CONTINUE_CODE;
            } else {
                printf("Command not recognized!\n");

                return CONTINUE_CODE;
            } /* default case for words array */
            words = {};
        } else {
            printf("No command given!\n");

            return CONTINUE_CODE;
        } /* END if (words.size() > 0) */
    } else {
        printf("No command given!\n");

        return CONTINUE_CODE;
    } /* END if (in.size() > 0) */

    // should never get here!
    return EXIT_FAILURE;
}

/**
 * @desc entry point for the ORPG program. This contains the main loop for the ORPG
 * shell. All command line arguments are parsed before entering the ORPG shell, and
 * the program may terminate before allowing user input.
 *
 * @param string in - the users input to be parsed
 * @return int - an integer code following the C/C++ standard for program success
 **/
int main(int argc, char* argv[]) {
    int status = parse_args(argc, argv); // may exit

    if(status == CONTINUE_CODE) {
        if(!Core::QUIET_FLAG) {
            // TODO - add more banners and randomly pick one
            Utils::print_file("banners/welcome_mat1");
        }

        string in("");

        // get user input
        while(status == CONTINUE_CODE) {
#           ifndef _WIN32
                printf("\33[4morpg\33[0m > ");
#           else
                // Windows does not currently support linux style terminal text editing
                printf("orpg > ");
#           endif

            Utils::safeGetline(cin, in);
            status = parse_input(in);
        }
    }
    return status;
}
